// esn/lib/html.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.appendScript = appendScript;
/**
 * 动态加载脚本文件
 */
function appendScript(src, text, reload, charset) {
    var id = hash(src + text);
    if (!reload && in_array(id, evalscripts)) return;
    if (reload && $(id)) {
        $(id).parentNode.removeChild($(id));
    }

    evalscripts.push(id);
    var scriptNode = document.createElement("script");
    scriptNode.type = "text/javascript";
    scriptNode.id = id;
    scriptNode.charset = charset ? charset : BROWSER.firefox ? document.characterSet : document.charset;
    try {
        if (src) {
            scriptNode.src = src;
            scriptNode.onloadDone = false;
            scriptNode.onload = function () {
                scriptNode.onloadDone = true;
                JSLOADED[src] = 1;
            };
            scriptNode.onreadystatechange = function () {
                if ((scriptNode.readyState == 'loaded' || scriptNode.readyState == 'complete') && !scriptNode.onloadDone) {
                    scriptNode.onloadDone = true;
                    JSLOADED[src] = 1;
                }
            };
        } else if (text) {
            scriptNode.text = text;
        }
        document.getElementsByTagName('head')[0].appendChild(scriptNode);
    } catch (e) {}
}

//下载类
//创建一个下载，output是地址，downloadFileName是下载的文件名
var downloads = exports.downloads = function downloads(output) {
    var downloadFileName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '文件';

    if (window.navigator.msSaveBlob) {
        // for ie 10 and later
        try {
            var blobObject = new Blob([output]);
            window.navigator.msSaveBlob(blobObject, downloadFileName);
        } catch (e) {
            console.log(e);
        }
    } else {
        var file = 'data:text/plain;charset=utf-8,';
        var logFile = output;
        var encoded = encodeURIComponent(logFile);
        file += encoded;
        var a = document.createElement('a');
        a.href = logFile;
        a.target = '_blank';
        a.download = downloadFileName;
        document.body.appendChild(a);
        a.click();
        a.remove();
    }
};

//web开发相关,获取url中的参数
var getUrlParam = exports.getUrlParam = function getUrlParam(name) {
    var reg = new RegExp('(^|&)' + name + '=([^&]*)(&|$)');
    var r = window.location.search.substr(1).match(reg);
    if (r != null) return decodeURI(unescape(r[2]));
    return '';
};

//获取站点地址
var getLocation = exports.getLocation = function getLocation() {
    return location.origin;
};

//判断当前平台是安卓还是ios或者其他
var platform = exports.platform = function platform() {
    var an = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : function () {};
    var ios = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : function () {};
    var any = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function () {};

    var u = navigator.userAgent,
        app = navigator.appVersion;
    var isAndroid = u.indexOf('Android') > -1 || u.indexOf('Linux') > -1; //g
    var isIOS = !!u.match(/\(i[^;]+;( U;)? CPU.+Mac OS X/); //ios终端
    if (isAndroid) {
        //这个是安卓操作系统
        an();
    } else if (isIOS) {
        //这个是ios操作系统
        ios();
    } else {
        any();
    }
};

//获取页面宽度
var getWidth = exports.getWidth = function getWidth() {
    return document.body.scrollWidth;
};

//对象转url字符串
var urlencode = exports.urlencode = function urlencode(str) {
    str = (str + '').toString();

    return encodeURIComponent(str).replace(/!/g, '%21').replace(/'/g, '%27').replace(/\(/g, '%28').replace(/\)/g, '%29').replace(/\*/g, '%2A').replace(/%20/g, '+');
};
