// esn/lib/storage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.cun = cun;
exports.qu = qu;
exports.quObj = quObj;
exports.shan = shan;
exports.cuns = cuns;
exports.qus = qus;
exports.qusObj = qusObj;
exports.shans = shans;
exports.shansAll = shansAll;
exports.cunc = cunc;
exports.quc = quc;
exports.shanc = shanc;
//localStorage的
function cun(key, value) {
    if (typeof value == 'string') {
        return localStorage.setItem(key, value);
    } else if (typeof value == "undefined") {
        return localStorage.setItem(key, '');
    } else if (typeof value == 'number') {
        return localStorage.setItem(key, value.toString());
    } else if ((typeof value === 'undefined' ? 'undefined' : _typeof(value)) == 'object') {
        return localStorage.setItem(key, JSON.stringify(value));
    }
}

function qu(val) {
    return localStorage.getItem(val);
}

function quObj(val) {
    var data = localStorage.getItem(val);
    if (data) {
        return JSON.parse(localStorage.getItem(val));
    } else {
        return null;
    }
}

function shan(val) {
    return localStorage.removeItem(val);
}

//sessionStorage的
function cuns(key, value) {
    if (typeof value == 'string') {
        return sessionStorage.setItem(key, value);
    } else if (typeof value == "undefined") {
        return sessionStorage.setItem(key, '');
    } else if (typeof value == 'number') {
        return sessionStorage.setItem(key, value.toString());
    } else if ((typeof value === 'undefined' ? 'undefined' : _typeof(value)) == 'object') {
        return sessionStorage.setItem(key, JSON.stringify(value));
    }
}

function qus(val) {
    return sessionStorage.getItem(val);
}

function qusObj(val) {
    var data = sessionStorage.getItem(val);
    if (data) {
        return JSON.parse(sessionStorage.getItem(val));
    } else {
        return null;
    }
}

function shans(val) {
    return sessionStorage.removeItem(val);
}

function shansAll() {
    return sessionStorage.clear();
}

//cookies的
//时间是天,默认30天
function cunc(name, value) {
    var time = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 30;
    var path = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : "/";

    if (!name || !value) return;
    var exp = new Date();
    exp.setTime(exp.getTime() + time * 24 * 60 * 60 * 1000);
    document.cookie = name + "=" + encodeURIComponent(value) + ";expires=" + exp.toUTCString() + ('; path=' + path + ';');
}

function quc(name) {
    var arr = void 0,
        reg = new RegExp("(^| )" + name + "=([^;]*)(;|$)");
    if (arr = document.cookie.match(reg)) return unescape(arr[2]);else return null;
}

//删除cookie
function shanc(name) {
    var exp = new Date();
    exp.setTime(exp.getTime() - 1);
    var cval = quc(name);
    if (!cval) document.cookie = name + "=" + cval + ";expires=" + exp.toUTCString();
}
